Sources/SwiftFusion/Inference/ChordalInitialization.swift

```diff
@@ -0,0 +1,251 @@
+// Copyright 2020 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import PenguinStructures
+import TensorFlow
+
+/// A relaxed version of the Rot3 between that uses the Chordal (Frobenious) norm on rotation
+/// Please refer to Carlone15icra (Initialization Techniques for 3D SLAM: a Survey on Rotation Estimation and its Use in Pose Graph Optimization)
+/// for explanation.
+public struct RelaxedRotationFactorRot3: LinearizableFactor
+{
+  public typealias Variables = Tuple2<Matrix3, Matrix3>
+  public typealias JacobianRows = Array9<Tuple2<Matrix3.TangentVector, Matrix3.TangentVector>>
```

Comment:
@marcrasi I am confused why it’s not ` Tuple2<array9<>>`. I definitely think of 2 9*9 jacobians, one for R1 and one for R2 argument.

Reply:
A few limitations are forcing it to be this type for now, but it should be possible to eventually have `Tuple2<Array9<>, Array9<>>`.

Why it's forced to be this type for now: The pullback of `errorVector` returns rows of the jacobian of `errorVector`. Since `errorVector` arity-2 in this case (or arity-N in the general case), the rows of the jacobian have type `Tuple2<A, B>` where `A` is the piece corresponding to the first argument and `B` is the piece corresponding to the second argument. When you combine all the rows together to get the whole jacobian matrix, you get `Array9<Tuple2<A, B>>`.

How we could change it to `Tuple2<Array9<>, Array9<>>`:
* Use forward mode ("differential") instead of reverse mode ("pullback") differentiation. This will give us columns of the jacobian, which can be combined together into a `Tuple2<Array9<>, Array9<>>`. But forward mode doesn't work well yet, so this approach is blocked for now.
* "Transpose" the `Array9<Tuple2<>>` that we get from the pullback. It's not trivial to express the type of the "transpose" method in Swift. I think it'll be possible once we develop the fixed-size linear algebra API a bit more.

Also note: with https://github.com/borglab/SwiftFusion/pull/104, this type will show up in fewer places (e.g. it'll go away from here).